File: elixir/lib/day8/day8.ex
defmodule Day8 do
  @moduledoc """
  Advent of Code 2024 - Day 8
  """

  def read_file(path) do
    {status, content} = File.read(path)

    if status != :ok do
      :break
    end

    get_dataset(String.replace(content, ~r/\r/, ""))
  end

  def run do
    dataset_test = read_file('lib/day8/test.txt')
    dataset = read_file('lib/day8/input.txt')

    {time_test1, result_test1} = :timer.tc(fn -> part1(dataset_test) end)
    {time1, result1} = :timer.tc(fn -> part1(dataset) end)
    {time_test2, result_test2} = :timer.tc(fn -> part2(dataset_test) end)
    {time2, result2} = :timer.tc(fn -> part2(dataset) end)

    IO.puts("#{IO.ANSI.light_white()}")
    IO.puts("#{IO.ANSI.red()}[DEMO] Part 1: #{time_format(time_test1)}s #{result_test1}")
    IO.puts("#{IO.ANSI.red()}[REAL] Part 1: #{time_format(time1)}s #{result1}")
    IO.puts("#{IO.ANSI.red()}[DEMO] Part 2: #{time_format(time_test2)}s #{result_test2}")
    IO.puts("#{IO.ANSI.red()}[REAL] Part 2: #{time_format(time2)}s #{result2}")
    IO.puts("#{IO.ANSI.light_white()}")
  end

  defp time_format(seconds) do
    :io_lib.format("~.6f", [seconds / 1_000_000]) |> List.to_string()
  end

  def get_dataset(content) do
    lines = String.split(content, "\n", trim: true)

      Enum.reduce(lines, [], fn line, n ->
        n ++ [String.graphemes(line)]
      end)
  end

  def combinations(list) do
    for x <- list, y <- list, x != y, do: [x, y]
  end

  def diff({x1, y1}, {x2, y2}) do
    {x1 - x2, y1 - y2}
  end

  def new_pos({x1, y1}, {x2, y2}) do
    {dx, dy} = diff({x1, y1}, {x2, y2})

    [
      {if x2 == dx + x1 do
         -1
       else
         dx + x1
       end,
       if y2 == dy + y1 do
         -1
       else
         dy + y1
       end},
      {if x1 == dx + x2 do
         -1
       else
         dx + x2
       end,
       if y1 == dy + y2 do
         -1
       else
         dy + y2
       end}
    ]
  end

  def process(acc_j, coord1, coord2, dataset) do
    Enum.reduce(new_pos(coord1, coord2), acc_j, fn {x, y}, acc_k ->
      if x >= 0 and x < length(dataset) and y >= 0 and y < length(Enum.at(dataset, 0)) do
        acc_k ++ [{x, y}]
      else
        acc_k
      end
    end)
  end

  def part1(dataset) do
    data =
      Enum.reduce(Enum.with_index(dataset), %{}, fn {row, i}, acc_i ->
        Enum.reduce(Enum.with_index(row), acc_i, fn {element, j}, acc_j ->
          if element != "." do
            Map.update(acc_j, element, [{i, j}], fn list -> [{i, j} | list] end)
          else
            acc_j
          end
        end)
      end)

    positions =
      Enum.reduce(data, [], fn {_key, coords}, acc_i ->
        Enum.reduce(combinations(coords), acc_i, fn [coord1, coord2], acc_j ->
          process(acc_j, coord1, coord2, dataset)
        end)
      end)

    Enum.uniq(positions) |> Enum.count()
  end

  def part2(_dataset) do
    0
  end
end

Day8.run()

File: php/src/A2024/Day8/Day8.php
<?php

namespace Factor\Aoc\A2024\Day8;

require __DIR__ . '/../../AdventOfCode.php';

use Factor\Aoc\AdventOfCode;

class Day8 extends AdventOfCode
{
    public function readDataset(): void
    {
        $dataset = [];

        $default = '............
........0...
.....0......
.......0....
....0.......
......A.....
............
............
........A...
.........A..
............
............';

        foreach (file($this->file) as $line) {
//        foreach (explode("\n", $default) as $line) {
            $dataset[] = str_split(trim(str_replace("\n", "", $line)));
        }

        $this->dataset = $dataset;
    }

    public function combinations(array $d): array
    {
        $r = [];
        foreach ($d as $a) {
            foreach ($d as $b) {
                if ($a !== $b) {
                    $r[] = [$a, $b];
                }
            }
        }

        return $r;
    }

    public function diff($a, $b): array
    {
        [$x1, $y1] = $a;
        [$x2, $y2] = $b;

        return [$x1 - $x2, $y1 - $y2];
    }

    public function new_pos($a,$b): array
    {
        [$x1, $y1] = $a;
        [$x2, $y2] = $b;

        return [$x1 + $x2 * 2, $y1 + $y2 * 2];
    }

    public function part1(): int
    {
        $coords = [];

        $width = count($this->dataset);
        $height = count($this->dataset[0]);

        for ($i = 0; $i < $width; $i++) {
            for ($j = 0; $j < $height; $j++) {
                if ($this->dataset[$i][$j] !== '.') {
                    $coords[$this->dataset[$i][$j]][] = [$i, $j];
                }
            }
        }

        $r = array_map(fn ($coord) => $this->combinations($coord), $coords);

        $s = $this->process($r, $width, $height);

        return count(array_unique($s));
    }

    public function process(array $r, int $width, int $height): array
    {
        $s = [];

        foreach ($r as $k => $coords) {
            foreach ($coords as $coord) {
                [$a, $b] = $coord;

                $n = $this->diff($a, $b);
                [$x, $y] = $this->new_pos($b,$n);

                if ($x >= 0 && $x < $width && $y >= 0 && $y < $height && $this->dataset[$x][$y] !== $k) {
                    $s[] = $x . '|' . $y;
                }
            }
        }

        return $s;
    }

    public function process2(array $r, int $width, int $height, &$s): array
    {
        foreach ($r as $k => $coords) {
            foreach ($coords as $coord) {
                [$a, $b] = $coord;

                $n = $this->diff($a, $b);
                [$x, $y] = $this->new_pos($b,$n);

                if ($x >= 0 && $x < $width && $y >= 0 && $y < $height && $this->dataset[$x][$y] !== $k) {
                    $s[] = $x . '|' . $y;
                }
            }
        }

        return $s;
    }


    public function part2(): int
    {
        $coords = [];

        $width = count($this->dataset);
        $height = count($this->dataset[0]);

        for ($i = 0; $i < $width; $i++) {
            for ($j = 0; $j < $height; $j++) {
                if ($this->dataset[$i][$j] !== '.') {
                    $coords[$this->dataset[$i][$j]][] = [$i, $j];
                }
            }
        }

        $r = array_map(fn ($coord) => $this->combinations($coord), $coords);

        $s = [];

        $this->process2($r, $width, $height, $s);

        return count(array_unique($s));
    }
}

$day = new Day8(__DIR__ . '/8.txt');
$day->run();
